### PR TITLE
x11: get current time via ICCCM recommendation

### PR DESCRIPTION
**Pull request by tych0**
Here's another attempt at fixing #1548

### CODE REVIEW

**Comment by zordsdavini** [comment]
I've tried. 1st impression was "whee, it works", but then I realized that I can't click by mouse or give any other Qtile keybinding command. In another scenario Qtile everything works until entering such app like Jetbrains. Will dig for some logs

File: libqtile/backend/x11/xcore.py [hunk 1]
```diff
@@ -328,7 +323,30 @@ def get_valid_timestamp(self):
...
+                0,
+                "",
+            )
+            while True:
+                event = conn.conn.wait_for_event()
+                if event.__class__ != xcffib.xproto.PropertyNotifyEvent:
+                    continue
+                return event.time
+        finally:
+            conn.finalize()
```
Comment (zordsdavini) [inline libqtile/backend/x11/xcore.py:349]: we don't have `conn` on exception

Comment (zordsdavini) [reply libqtile/backend/x11/xcore.py:349]: maybe we should log exception for better understanding?

**Comment by zordsdavini** [comment]
```
2020-05-11 06:43:02,589 ERROR libqtile xcore.py:_xpoll():L278 Got an exception in poll loop
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/libqtile/backend/x11/xcore.py", line 333, in get_valid_timestamp
    conn = xcbq.Connection(display_name)
NameError: name 'display_name' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/libqtile/backend/x11/xcore.py", line 246, in _xpoll
    ret = target(event)
  File "/usr/lib/python3.8/site-packages/libqtile/backend/x11/xcore.py", line 575, in handle_MapRequest
    self.qtile.map_window(window)
  File "/usr/lib/python3.8/site-packages/libqtile/core/manager.py", line 522, in map_window
    c = self.manage(window)
  File "/usr/lib/python3.8/site-packages/libqtile/core/manager.py", line 575, in manage
    self.current_screen.group.add(c, focus=c.can_steal_focus())
  File "/usr/lib/python3.8/site-packages/libqtile/group.py", line 271, in add
    self.focus(win, warp=True, force=force)
  File "/usr/lib/python3.8/site-packages/libqtile/group.py", line 233, in focus
    layout.focus(win)
  File "/usr/lib/python3.8/site-packages/libqtile/layout/base.py", line 604, in focus
    self.group.layout_all()
  File "/usr/lib/python3.8/site-packages/libqtile/group.py", line 167, in layout_all
    self.current_window.focus(warp)
  File "/usr/lib/python3.8/site-packages/libqtile/window.py", line 559, in focus
    self.qtile.core.get_valid_timestamp(),
  File "/usr/lib/python3.8/site-packages/libqtile/backend/x11/xcore.py", line 349, in get_valid_timestamp
    conn.finalize()
UnboundLocalError: local variable 'conn' referenced before assignment
```

**Comment by pawamoy** [comment]
Same as @zordsdavini: `display_name` and `conn` not defined or referenced before assignment.

Also, worse than ever, once a PyCharm window was open, I could not unfocus it anymore.

- keyboard shortcuts for Qtile: nothing happens
- clicking on the top bar groups: nothing happens (cursor frozen into "edition" mode)
- clicking in the PyCharm window: nothing happens

The only thing possible was writing in the current opended file in PyCharm. PyCharm shortcuts were working as well, unfortunately I did not know the one to close PyCharm haha

**Comment by tych0** [comment]
Oof, sorry. I had hoped the test suite would exercise this code, but apparently it doesn't. I fixed the two obvious bugs, any chance you guys can try again?

**Comment by zordsdavini** [comment]
So, did tests. Same as before (Qtile hangs, no mouse clicks are working) + no these errors :rofl: 
I have logs that maybe are related, but I think I had them before:
```
2020-05-11 17:08:57,998 WARNING libqtile xcbq.py:get_property():L715 X error in GetProperty (wid=46137345, prop='WM_PRO
TOCOLS'), ignoring
2020-05-11 17:08:58,000 WARNING libqtile xcbq.py:set_property():L685 X error in SetProperty (wid=46137345, prop='WM_STA
TE'), ignoring
2020-05-11 17:12:09,500 WARNING libqtile xcbq.py:set_property():L685 X error in SetProperty (wid=20977617, prop='WM_STA
TE'), ignoring
2020-05-11 17:15:18,324 WARNING libqtile xcbq.py:set_property():L685 X error in SetProperty (wid=20986082, prop='WM_STA
TE'), ignoring
2020-05-11 17:18:01,144 WARNING libqtile xcbq.py:set_property():L685 X error in SetProperty (wid=56623109, prop='WM_STA
TE'), ignoring
2020-05-11 17:21:22,700 ERROR libqtile xcore.py:_xpoll():L280 Got an exception in poll loop
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/libqtile/backend/x11/xcore.py", line 248, in _xpoll
    ret = target(event)
  File "/usr/lib/python3.8/site-packages/libqtile/backend/x11/xcore.py", line 497, in handle_KeyPress
    self.qtile.process_key_event(keysym, event.state & self._valid_mask)
  File "/usr/lib/python3.8/site-packages/libqtile/core/manager.py", line 362, in process_key_event
    key = self.keys_map[(keysym, mask)]
KeyError: (2770, 64)
```

**Comment by tych0** [comment]
Huh, qtile hangs. Can you put some logs in and see exactly where? Seems like it's probably stuck polling for the event maybe?

**Comment by tych0** [comment]
Ok, one last idea, then I probably need to figure out how to reproduce this locally so I can debug it. See if the above patch helps?

**Comment by zordsdavini** [comment]
Nope, didn't help. Same logs as above :(

**Comment by tych0** [comment]
Ok. Can you confirm where it's hanging?

**Comment by zordsdavini** [comment]
to test locally - you can install pycharm, phpstorm, datagrip or any other jetbrains product locally, open file to edit and click other window (or group) and back - no cursor

**Comment by zordsdavini** [comment]
yes

**Comment by tych0** [comment]
On Mon, May 11, 2020 at 08:27:09AM -0700, Arnas Udovičius wrote:
> yes

Ok, where is it? :)

**Comment by zordsdavini** [comment]
What where? Qtile not responding to key press or mouse clicks, no focus on other client, so I can only kill Qtile to get it working again. If I focus on Jetbrains app - Qtile stops responding again

**Comment by tych0** [comment]
> Can you confirm where it's hanging?

i.e. with some logs, can you shed some light on exactly where things are going wrong?

**Comment by zordsdavini** [comment]
Sorry I read it wrong. No, aall logs that I gave before. Maybe I should turn on debug. Will try tomorrow

**Comment by tych0** [comment]
> No, aall logs that I gave before. Maybe I should turn on debug. Will try tomorrow

I'm suggesting you add some of your own logs :)

File: libqtile/backend/x11/xcore.py [hunk 2]
```diff
@@ -328,7 +323,35 @@ def get_valid_timestamp(self):
...
+                xcffib.xproto.PropMode.Append,
+                self._root.wid,
+                self.conn.atoms["WM_CLASS"],
+                self.conn.atoms["STRING"],
+                8,
+                0,
+                "",
+            )
+            while True:
+                event = conn.conn.wait_for_event()
```
Comment (zordsdavini) [inline libqtile/backend/x11/xcore.py:348]: added some debug messages and Qtile hangs in this line after starting Jetbrains app

**Comment by zordsdavini** [comment]
I just set `return xcffib.xproto.Time.CurrentTime` in `get_valid_timestamp` and it looks like it works much better. But this is revert of all changes from https://github.com/qtile/qtile/commit/28d5c30005b5342db9ea400e638daad15d110bfe , so IDK is it should go in that way

**Comment by tych0** [comment]
Yes, I think we should not revert it, we should just get this fix right. Then it will work for everyone :)

File: libqtile/backend/x11/xcore.py (hunk 2 above)
Comment (tych0) [reply libqtile/backend/x11/xcore.py:348]: Ok, thanks. Seems like something about the masking is wrong :(

**Comment by tych0** [comment]
Ok, seems like this was maybe just my silly mistake. Anyone using these IDEs have time to try it now?

**Comment by pawamoy** [comment]
Yes, seems to be working fine :slightly_smiling_face: ! I still have trouble getting the focus on the editor part of the window, but none of the previous critical issues :+1:

**Comment by tych0** [comment]
So does just using CurrentTime instead of `get_valid_timestamp()` fix all the issues for you? Or is it that we've fixed that problem but have introduced other regressions?

**Comment by pawamoy** [comment]
It made JetBrains products partially usable for 0.15. They were already partially usable in 0.14, but not usable at all in 0.15 without the changes in this PR.

However there are still issues with focus (that were already here in 0.14). They're just a bit different than before, and bit more annoying.

**Comment by tych0** [comment]
If you take out all the java hacks, does that help? Maybe they fixed their behavior and we can just drop them all?

**Comment by pawamoy** [comment]
I tried this:

- `wmname = "LG3D"` without `export _JAVA_AWT_WM_NONREPARENTING=1`: works, focus/input issues
- `wmname = "LG3D"` with `export _JAVA_AWT_WM_NONREPARENTING=1`: works, focus/input issues
- `wmname = "qtile"` without `export _JAVA_AWT_WM_NONREPARENTING=1`: doesn't work
- `wmname = "qtile"` with `export _JAVA_AWT_WM_NONREPARENTING=1`: works, focus/input issues

Not sure what other Java hacks you could be talking about :slightly_smiling_face:

**Comment by tych0** [comment]
Sorry, should have mentioned that!

https://github.com/qtile/qtile/blob/master/libqtile/window.py#L543

is what I was thinking of.

**Comment by pawamoy** [comment]
Changed it to 

```
is_java_main = False
is_java_dialog = False
is_java = is_java_main or is_java_dialog
```

and didn't notice any difference.

**Comment by zordsdavini** [comment]
Did some tests, too
Yes, it looks like this fixed issues what has been discovered in 0.15 version, so we should merge it.
Next, old issue with loosing focus but easily getting it back (ex., switching between internal tabs) still persists.
In debug it looks like:
```
2020-05-18 07:56:41,029 libqtile hook.py:fire():L353  Internal event: client_focus((Window('phpstorm_prj – Project'),), {})
2020-05-18 07:56:41,030 libqtile hook.py:fire():L353  Internal event: focus_change((), {})
2020-05-18 07:56:41,052 libqtile hook.py:fire():L353  Internal event: client_focus((Window('phpstorm_prj – Project'),), {})
```
Here we loos focus for carriet after swithing to other client (terminal in my case) and back

**Comment by TauPan** [comment]
Interesting.

With this PR I can focus on an input field in pyCharm by clicking twice with the left mouse.

And when I click into the program text, apparently it has focus, but I see no caret.

**Comment by tych0** [comment]
Ok, I'm going to merge this, since it seems like a strict improvement over what's in the tree now. Seems like we still have some focus issues, though. I looked around a bit and found:

https://github.com/tych0/qtile/commit/a38454255fcd4abb465984fccab74b297f244d5a

Can anyone test and see if that helps?

**Comment by zordsdavini** [comment]
with your merged diff + `elif` I got input all the time. For me it looks awesome. Even better then in 0.14.2 version

**Comment by tych0** [comment]
Cool, thanks! I sent the official PR here:

https://github.com/qtile/qtile/pull/1736

**Comment by jsdima** [comment]
I also confirm, with tych0@a384542 works perfect, always focused and the cursor stays in place. Great job, thanks.

**Pull request merged**